### PR TITLE
Added gas fee strategy

### PR DESCRIPTION
**Pull request by kantagara**
so developers can specify which exact strategy they wanna use for their gas fee calculation
#1110

### CODE REVIEW

**Review by rob1997 — APPROVED** [review]

File: src/UnitySampleProject/Assets/Samples/web3.unity SDK/2.6.0/Web3.Unity Samples/Scenes/SampleLogin - Connect Modal.unity [hunk 1]
```diff
@@ -215,6 +215,53 @@ Transform:
   m_Children: []
   m_Father: {fileID: 0}
   m_LocalEulerAnglesHint: {x: 0, y: 0, z: 0}
+--- !u!1 &1270439810
```
Comment (rob1997) [inline src/UnitySampleProject/Assets/Samples/web3.unity SDK/2.6.0/Web3.Unity Samples/Scenes/SampleLogin - Connect Modal.unity:218]: is this change necessary, I ask because it'll cause a merge conflict with my PRs

File: src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs [hunk 2]
```diff
@@ -27,10 +28,20 @@ public WalletTransactionExecutor(IWalletProvider walletProvider, ISigner signer)
             this.signer = signer;
         }
 
-        public async Task<TransactionResponse> SendTransaction(TransactionRequest transaction)
+        public async Task<TransactionResponse> SendTransaction(TransactionRequest transaction, IGasFeeModifier gasFeeModifier = null)
         {
             transaction.From ??= signer.PublicAddress;
             transaction.Data ??= "0x";
+            gasFeeModifier ??= new NoGasFeeModifier();
```
Comment (rob1997) [inline src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs:35]: Idk if we need this, wallet handle these independently I think

Comment (kantagara) [reply src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs:35]: Ahhhh you are right, but then gasFeeModifier will not be used leading to a nullref down below. I was banging my head on that 😆

Comment (oleksandrchainsafe) [reply src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs:35]: I'm not an expert, but I think wallets do pick the gas price for you. Some of them (all?) even allow to set the gas price yourself. If it's the case, this functionality should still be useful for presetting gas fee strategies by developers for specific transactions, I think.

To be more confident we could just test out how this works with different wallets.

Another question is where does the 'initial' value of `transaction.GasPrice` and `transaction.MaxFeePerGas` comes from. Does we calculate/fetch it ourself somewhere?

File: src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs [hunk 3]
```diff
@@ -27,10 +28,20 @@ public WalletTransactionExecutor(IWalletProvider walletProvider, ISigner signer)
...
+
+            if (!transaction.GasPrice.Value.IsZero)
+            {
+                transaction.GasPrice = gasFeeModifier.ModifyGasFee(transaction.GasPrice);
+            }
+            else if(!transaction.MaxFeePerGas.Value.IsZero)
+            {
+                transaction.GasPrice = gasFeeModifier.ModifyGasFee(transaction.GasPrice);
```
Comment (oleksandrchainsafe) [inline src/ChainSafe.Gaming/Web3/Evm/Wallet/WalletTransactionExecutor.cs:43]: this should probably use `transaction.MaxFeePerGas` not `transaction.GasPrice`, right?

**Review by oleksandrchainsafe — CHANGES_REQUESTED** [review]

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
Im getting 2 errors when I click play in the editor:

```
System.IO.IOException: The file '/Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnitySampleProject/Assets/Resources/HyperPlayConnectionProvider.asset' already exists.
  at System.IO.FileSystem.LinkOrCopyFile (System.String sourceFullPath, System.String destFullPath) [0x000b9] in <e307bbb467104258887a104f6151f183>:0 
  at System.IO.FileSystem.MoveFile (System.String sourceFullPath, System.String destFullPath) [0x0003a] in <e307bbb467104258887a104f6151f183>:0 
  at System.IO.File.Move (System.String sourceFileName, System.String destFileName) [0x00083] in <e307bbb467104258887a104f6151f183>:0 
  at ChainSafe.GamingSdk.Editor.ReadWriteProvider.TryMovingConnectionProviders () [0x0006e] in /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Editor/ReadWriteProvider.cs:38 
  at ChainSafe.GamingSdk.Editor.ReadWriteProvider..cctor () [0x00000] in /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Editor/ReadWriteProvider.cs:17 
UnityEditor.EditorAssemblies:ProcessInitializeOnLoadAttributes (System.Type[]) (at /Users/bokken/build/output/unity/unity/Editor/Mono/EditorAssemblies.cs:127)

ArgumentNullException: Server URL is required.
Plugins.CountlySDK.Countly.Init (Plugins.CountlySDK.Models.CountlyConfiguration configuration) (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Plugins/Countly SDK/Plugins/CountlySDK/Countly.cs:331)
Plugins.CountlySDK.Countly.Awake () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Plugins/Countly SDK/Plugins/CountlySDK/Countly.cs:152)
```

**Comment by juans-chainsafe** [comment]
> Im getting 2 errors when I click play in the editor:
> 
> ```
> System.IO.IOException: The file '/Users/juanmanuelspoleti/Desktop/workspace/web3.unity/src/UnitySampleProject/Assets/Resources/HyperPlayConnectionProvider.asset' already exists.
>   at System.IO.FileSystem.LinkOrCopyFile (System.String sourceFullPath, System.String destFullPath) [0x000b9] in <e307bbb467104258887a104f6151f183>:0 
>   at System.IO.FileSystem.MoveFile (System.String sourceFullPath, System.String destFullPath) [0x0003a] in <e307bbb467104258887a104f6151f183>:0 
>   at System.IO.File.Move (System.String sourceFileName, System.String destFileName) [0x00083] in <e307bbb467104258887a104f6151f183>:0 
>   at ChainSafe.GamingSdk.Editor.ReadWriteProvider.TryMovingConnectionProviders () [0x0006e] in /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Editor/ReadWriteProvider.cs:38 
>   at ChainSafe.GamingSdk.Editor.ReadWriteProvider..cctor () [0x00000] in /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Editor/ReadWriteProvider.cs:17 
> UnityEditor.EditorAssemblies:ProcessInitializeOnLoadAttributes (System.Type[]) (at /Users/bokken/build/output/unity/unity/Editor/Mono/EditorAssemblies.cs:127)
> 
> ArgumentNullException: Server URL is required.
> Plugins.CountlySDK.Countly.Init (Plugins.CountlySDK.Models.CountlyConfiguration configuration) (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Plugins/Countly SDK/Plugins/CountlySDK/Countly.cs:331)
> Plugins.CountlySDK.Countly.Awake () (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Plugins/Countly SDK/Plugins/CountlySDK/Countly.cs:152)
> ```

this was solved! Now the issue is that the new parameter is mandatory, so it fails if you dont put it:
```
NullReferenceException: Object reference not set to an instance of an object
ChainSafe.Gaming.Web3.Evm.Wallet.WalletTransactionExecutor.SendTransaction (ChainSafe.Gaming.Evm.Transactions.TransactionRequest transaction, ChainSafe.Gaming.Evm.Contracts.GasFees.IGasFeeModifier gasFeeModifier) (at <e2753505c7f6434b871e63ae6489ffa3>:0)
ChainSafe.Gaming.Evm.Contracts.Contract.SendWithReceipt (System.String method, System.Object[] parameters, ChainSafe.Gaming.Evm.Transactions.TransactionRequest overwrite, ChainSafe.Gaming.Evm.Contracts.GasFees.IGasFeeModifier gasFeeModifier) (at <e2753505c7f6434b871e63ae6489ffa3>:0)
ChainSafe.Gaming.Evm.Contracts.Contract.Send (System.String method, System.Object[] parameters, ChainSafe.Gaming.Evm.Transactions.TransactionRequest overwrite, ChainSafe.Gaming.Evm.Contracts.GasFees.IGasFeeModifier gasFeeModifier) (at <e2753505c7f6434b871e63ae6489ffa3>:0)
ChainSafe.Gaming.Evm.Contracts.BuiltIn.Erc721Contract.Mint (System.String uri, ChainSafe.Gaming.Evm.Contracts.GasFees.IGasFeeModifier gasFeeModifier) (at <e2753505c7f6434b871e63ae6489ffa3>:0)
Erc721Calls.MintErc721 () (at Assets/Samples/web3.unity SDK/2.6.0/Web3.Unity Samples/Scripts/Scenes/SampleMain/Erc721/Erc721Calls.cs:114)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <e307bbb467104258887a104f6151f183>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

**Comment by kantagara** [comment]
Closing the PR since I am a dumb person. Will create a new one with proper implementation :D
